### ISSUE TITLE
Category

### ISSUE THREAD

**Issue by dimecon** [dimecon/free-cats#7]
For the equivalence class, I'm stuck on composition:

```
  eqClass : ∀ {a a′} → ExpCategory a a′ → FreeCategory a a′
  eqClass (Embed cf) = cf ∷ nil
  eqClass Id = nil
  eqClass (f ⟫ g) = {!!}

```

Unless I'm missing something, I think we need something like

```
  comp-of-exp : ∀ {a a′} → ExpCategory a a′ → C a a′
```

Otherwise I'm not sure how to utilize

```
    _∷_ : ∀ {a a′ a′′} → C a a′ → FreeCategory a′ a′′ → FreeCategory a  a′′
```

**Comment by gelisam** [comment]
It would not be possible to implement `comp-of-exp`, since an `ExpCategory` can contain more than one `C` and we don't have any way to combine two `C`s into one. We could ask the user to provide us with such an operation, but then we would no longer be providing a free Category! "Free" means that we start with a completely generic base type, so a `C` with no known operations, and we construct a new type around it which does support the Category operations. The Monoid analogy would be: we're able to construct the type `List C` out of any `C`, we don't require a method to combine two `C`s together.

So how does `List C` manage to support `mappend` even though `mappend`'s left argument is a `List C` while `_::_`'s left argument is a `C`? By implementing concatenation! It's the same thing here, we need to implement concatenation for `FreeCategory`.

**Comment by dimecon** [comment]
Ah, I see. How is this?

```
  _++_ : ∀ {a a′ a′′} → FreeCategory a a′ → FreeCategory a′ a′′ → FreeCategory a a′′
  nil ++ ys = ys
  (x ∷ xs) ++ ys = x ∷ (xs ++ ys)

  exp→free : ∀ {a a′} → ExpCategory a a′ → FreeCategory a a′
  exp→free (Embed cf) = cf ∷ nil
  exp→free Id = nil
  exp→free (f • g) = exp→free f ++ exp→free g
```

It looks like the soundness and completeness proofs will be pretty much exactly the same as the ones for `FreeMonoid`. Is that right?

**Comment by gelisam** [comment]
Yes, exactly! The remaining proofs are indeed exactly the same: Category is completely analogous to Monoid, and Semigroupoid is exactly analogous to Semigroup. The real challenge begins with the next level, Bifunctorial, because there is no Monoid-like equivalent.

**Comment by dimecon** [comment]
Instead of just copying the proofs over to category, do you know if there is a better way (such as generalizing the proofs so that category and monoid are just instances of it)?

**Comment by gelisam** [comment]
Here too, I think the category proofs will be general enough to cover the monoid case.
